security-oauth2/src/main/java/io/micronaut/security/oauth2/client/clientcredentials/ClientCredentialsHttpClientFilter.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.security.oauth2.client.clientcredentials;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import io.micronaut.context.BeanContext;
+import io.micronaut.context.exceptions.NoSuchBeanException;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.HttpResponse;
+import io.micronaut.http.MutableHttpRequest;
+import io.micronaut.http.annotation.Filter;
+import io.micronaut.http.filter.ClientFilterChain;
+import io.micronaut.http.filter.HttpClientFilter;
+import io.micronaut.http.util.OutgoingHttpRequestProcessor;
+import io.micronaut.inject.qualifiers.Qualifiers;
+import io.micronaut.security.oauth2.configuration.OauthClientConfiguration;
+import io.micronaut.security.oauth2.endpoint.token.response.TokenResponse;
+import io.micronaut.security.token.propagation.HttpHeaderTokenPropagator;
+import io.micronaut.security.token.propagation.TokenPropagator;
+import io.reactivex.Flowable;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * An {@link HttpClientFilter} to add an access token to outgoing request thanks to a  Client Credentials request.
+ *
+ * @author Sergio del Amo
+ * @since 2.2.0
+ */
+@Filter("/**")
+public class ClientCredentialsHttpClientFilter implements HttpClientFilter {
+    private static final Logger LOG = LoggerFactory.getLogger(ClientCredentialsHttpClientFilter.class);
+    protected final OutgoingHttpRequestProcessor outgoingHttpRequestProcessor;
+    protected final Collection<OauthClientConfiguration> oauthClientConfigurationCollection;
+    protected final TokenPropagator tokenPropagator;
+    protected final BeanContext beanContext;
+    protected final Map<String, ClientCredentialsClient> clientCredentialsClientsByName = new ConcurrentHashMap<>();
```

Comment:
This needs to be thread safe

Reply:
```java
if (!clientCredentialsClientsByName.containsKey(oauthClient.getName())) {
    try {
        ClientCredentialsClient clientCredentialsClient = beanContext.getBean(ClientCredentialsClient.class, Qualifiers.byName(oauthClient.getName()));
        clientCredentialsClientsByName.put(oauthClient.getName(), clientCredentialsClient);
    } catch (NoSuchBeanException e) {
        if (LOG.isTraceEnabled()) {
            LOG.trace("no client credentials client for OAuth 2.0 client {}", oauthClient.getName());
        }
    }
}
return Optional.ofNullable(clientCredentialsClientsByName.get(oauthClient.getName()));

```

with:

```java
try {
    clientCredentialsClientsByName.putIfAbsent(oauthClient.getName(), beanContext.getBean(ClientCredentialsClient.class, Qualifiers.byName(oauthClient.getName())));
    return Optional.of(clientCredentialsClientsByName.get(oauthClient.getName()));
} catch (NoSuchBeanException e) {
    if (LOG.isTraceEnabled()) {
        LOG.trace("no client credentials client for OAuth 2.0 client {}", oauthClient.getName());
    }
}
return Optional.empty();

```

Reply:
That is still not good because that means the `getBean` call will happen every time and multiple puts could occur

Reply:
I've replaced the usage of `putIfAbsent` with `computeIfAbsent`